work_queue.cpp

```diff
@@ -0,0 +1,89 @@
+//
+// Created by victor on 6/2/21.
+//
+
+#include "include/work_queue.h"
+
+const unsigned first_idx = 2;
+
+WorkQueue::WorkQueue(uint32_t buffer_size, Node nodes) : buffer_size
+(buffer_size), cq(20,buffer_size*sizeof(Node)), buffers(nodes) {
+  for (Node i = 0; i < nodes; ++i) {
+    buffers[i] = static_cast<unsigned long *>(malloc(buffer_size * sizeof(Node)));
+    buffers[i][0] = first_idx; // first spot will point to the next free space
+    buffers[i][1] = i; // second spot identifies the node to which the buffer
+    // belongs
+  }
+}
+
+WorkQueue::~WorkQueue() {
+  for (auto & buffer : buffers) {
+    free(buffer);
+  }
+}
+
+void WorkQueue::flush(Node *buffer, uint32_t num_bytes) {
+  cq.push(reinterpret_cast<char *>(buffer), num_bytes);
+}
+
+insert_ret_t WorkQueue::insert(update_t upd) {
+  Node& idx = buffers[upd.first][0];
+  buffers[upd.first][idx] = upd.second;
+  ++idx;
```

Comment:
This code assumes that it knows how big a Node is. The buffers array is of type unsigned long * which means this code only works if type(Node) = unsigned long.

This might not be a big problem but it could cause us problems if we ever want to change the type of Node

Reply:
I don't think so. The buffers array is of type `Node*`, and it will change if we change the type of `Node`.

Reply:
Okay, can we change line 12 to also be a static cast to Node* then?

Reply:
oops